client/scripts/views/components/autocomplete_tag_form.ts

```diff
@@ -1,121 +1,100 @@
+    };
 
-    const clearAutoComplete = (inputEle) => {
-      inputEle.value = '';
-      vnode.state.noMatches = false;
-      (document.getElementsByClassName('results')[0] as HTMLInputElement)
-        .classList.remove('show');
-      (document.getElementsByClassName('autocomplete-wrap')[0] as HTMLInputElement)
-        .classList.remove('displaying-results');
-      m.redraw();
-      inputEle.focus();
+    const onSelect = (item: OffchainTag) => {
+      vnode.state.selectedTag = item;
+      updateFormData(item.name, item.id);
     };
 
-    const addTag = (tag: string) => {
-      if (!tag.length) return;
-      const { tags } = vnode.state;
-      const { updateFormData } = vnode.attrs;
-      const formInput = (document.getElementsByClassName('autocomplete-entry')[0] as HTMLInputElement);
-      if (tags.length === 1) {
-        clearAutoComplete(formInput);
-        vnode.attrs.updateParentErrors('You can only select a single tag');
-        return;
-      }
-      for (const t of tags) {
-        if (t.toLowerCase() === tag.toLowerCase()) {
-          clearAutoComplete(formInput);
-          return;
-        }
-      }
-      vnode.attrs.updateParentErrors();
-      vnode.state.tags.push(tag);
-      updateFormData(vnode.state.tags);
-      clearAutoComplete(formInput);
+    const manuallyClosePopover = () => {
+      const button = document.getElementsByClassName('tag-selection-drop-menu')[0];
+      if (button) (button as HTMLButtonElement).click();
     };
 
-    const addTagWrap = (e) => {
-      e.preventDefault();
-      const formInput = (document.getElementsByClassName('autocomplete-entry')[0] as HTMLInputElement);
-      const tag = vnode.state.noMatches ? formInput.value : e.target.innerText;
-      addTag(tag);
+    const addTag = () => {
+      const input = (document.getElementsByClassName('autocomplete-tag-input')[0].firstChild as HTMLInputElement);
+      const newTag = input.value;
+      tags.push({ name: newTag, id: undefined, description: '' });
+      setTimeout(() => { vnode.state.selectedTag = newTag; }, 1);
+      updateFormData(newTag);
+      manuallyClosePopover();
     };
 
-    const rowComponentFunc = (res) => TagItem(res.name, addTagWrap, 'addTag');
+    const sortTags = (tags_: OffchainTag[]) => {
+      return tags_.filter((tag) => featuredTags.includes(tag)).sort((a, b) => a.name > b.name ? 1 : -1)
+        .concat(tags_.filter((tag) => !featuredTags.includes(tag)).sort((a, b) => a.name > b.name ? 1 : -1));
+    };
 
-    const onChangeHandler = (result) => {
-      if (!result.length) return false;
-      delete vnode.state.error;
-      if (result.length === 1) {
-        const formInput = (document.getElementsByClassName('autocomplete-entry')[0] as HTMLInputElement);
-        formInput.addEventListener('keyup', (e) => {
-          if (e.keyCode === 13) addTag(formInput.value);
-        });
+    const EmptyContent: m.Component<{}, {}> = {
```

Comment:
Can we move some of these functions/components outside of the higher-order component? They don't look like they depend on any internal state and there's a lot of them.